log4j-api-test/src/test/java/org/apache/logging/log4j/message/ParameterFormatterTest.java

```diff
@@ -17,156 +17,167 @@
+            final int placeholderCount,
+            final String placeholderCharIndicesString,
+            final boolean escapedPlaceholderFound,
+            final String pattern) {
+        ParameterFormatter.MessagePatternAnalysis analysis = ParameterFormatter.analyzePattern(pattern);
+        assertThat(analysis.placeholderCount).isEqualTo(placeholderCount);
+        if (placeholderCount > 0) {
+            final int[] placeholderCharIndices = Arrays
+                    .stream(placeholderCharIndicesString.split(":"))
+                    .mapToInt(Integer::parseInt)
+                    .toArray();
+            assertThat(analysis.placeholderCharIndices).startsWith(placeholderCharIndices);
+            assertThat(analysis.escapedCharFound).isEqualTo(escapedPlaceholderFound);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("messageFormattingTestCases")
+    void assertMessageFormatting(
+            final String pattern,
+            final Object[] args,
+            final int argCount,
+            final String expectedFormattedMessage) {
+        ParameterFormatter.MessagePatternAnalysis analysis = ParameterFormatter.analyzePattern(pattern);
+        final StringBuilder buffer = new StringBuilder();
+        ParameterFormatter.formatMessage(buffer, pattern, args, argCount, analysis);
+        String actualFormattedMessage = buffer.toString();
+        assertThat(actualFormattedMessage).isEqualTo(expectedFormattedMessage);
+    }
+
+    static Object[][] messageFormattingTestCases() {
+        return new Object[][]{
+                new Object[]{
+                        "Test message {}{} {}",
+                        new Object[]{"a", "b", "c"},
+                        3,
+                        "Test message ab c"
+                },
+                new Object[]{
+                        "Test message {} {} {} {} {} {}",
+                        new Object[]{"a", null, "c", null, null, null},
+                        6,
+                        "Test message a null c null null null"
+                },
+                new Object[]{
+                        "Test message {}{} {}",
+                        new Object[]{"a", "b", "c", "unnecessary", "superfluous"},
+                        5,
+                        "Test message ab c"
+                },
+                new Object[]{
+                        "Test message \\{}{} {}",
+                        new Object[]{"a", "b", "c"},
+                        3,
+                        "Test message {}a b"
+                },
+                new Object[]{
+                        "Test message {}{} {}\\",
+                        new Object[]{"a", "b", "c"},
+                        3,
+                        "Test message ab c\\"
+                },
+                new Object[]{
+                        "Test message {}{} {}\\\\",
+                        new Object[]{"a", "b", "c"},
+                        3,
+                        "Test message ab c\\"
```

Comment:
@rgoers, behaviour change: previously this was returning `Test message ab c\\\\`. I think this was a bug in the earlier version, since in other occasions `\\\\\` gets formatted to `\\`; for instance, see the next test scenario. Hence, this behaviour should be consistent, IMO and that is what I strived to capture in the new version.

Reply:
It has been two weeks. I am resolving this conversation.